blockchain/blockchain.go

```diff
@@ -859,15 +862,15 @@ func (b *Blockchain) executeBlockTransactions(block *types.Block) (*BlockResult,
 
 // WriteBlock writes a single block to the local blockchain.
 // It doesn't do any kind of verification, only commits the block to the DB
-func (b *Blockchain) WriteBlock(block *types.Block) error {
-	// Log the information
-	b.logger.Info(
-		"write block",
-		"num",
-		block.Number(),
-		"parent",
-		block.ParentHash(),
-	)
+func (b *Blockchain) WriteBlock(block *types.Block, source string) error {
+	b.writeLock.Lock()
+	defer b.writeLock.Unlock()
+
+	if block.Number() <= b.Header().Number {
+		b.logger.Info("block already inserted", "block", block.Number(), "source", source)
+
+		return nil
+	}
```

Comment:
We've discussed this internally, we need this logic here because the syncer and consensus modules concurrently write blocks now, and we need this safeguard in place.

Restore is not affected

Reply:
As discussed, we have decided to do that in the future task